modules/gmake2/gmake2_cpp.lua

```diff
@@ -60,7 +60,13 @@
 			fileExtension { ".cc", ".cpp", ".cxx", ".mm" }
 			buildoutputs  { "$(OBJDIR)/%{file.objname}.o" }
 			buildmessage  '$(notdir $<)'
-			buildcommands {'$(CXX) %{premake.modules.gmake2.cpp.fileFlags(cfg, file)} $(FORCE_INCLUDE) -o "$@" -MF "$(@:%.o=%.d)" -c "$<"'}
+			buildcommands {
+				-- code checking must be before the build, so that if the code checking 
+				-- fails, the build is not completed, and hence, a re-run will trigger 
+				-- the code check again.
+				'$(CLANG_TIDY) "$<" -- %{premake.modules.gmake2.cpp.fileFlags(cfg, file)} $(FORCE_INCLUDE) -o "$@" -MF "$(@:%.o=%.d)" -c "$<"',
```

Comment:
How difficult would it be to make it so this isn't emitted when clang-tidy isn't requested, rather than always emitting and just "ignoring" it sometimes?

Reply:
@nickclark2016 : I knew someone would ask this question, so I already pre-empted the answer in my original PR description: Unfortunately, at the point where the rules are decided, we do not have access to the value of the clangtidy property. You can see it here in this function, there is no information to play with. And then we have other issues as well like making these rules configuration specific.

Then again, I am an outsider to the premake team, you probably know more about the code and can suggest better.

Reply:
From "`rule`" point of view, we would need
- 2 rules: `rule 'cpp'` and `rule 'cpp_with_clang_tidy'`
- remove the `fileExtensions` and do the dispatch "manually".

Reply:
Could you please give some psuedo code to explain what you are saying? I'm not sure I understand.

Reply:
I mix with customcommand as fileextension cannot be removed...
Idea is to customize rule to handle clangtidy:
if you read https://premake.github.io/docs/Custom-Rules/
we might add `propertydefinition` to handle clangtidy, then we have to activate that property for each `filecfg` with clangtidy.
Not sure at which point it is doable without hack.
I will try to add clang-tidy support to premake-ninja, but it doesn't use (premake) `rule` internally but has "intrinsic" rule.
Above (adapted) method should do the trick for me.

Reply:
I see. This sounds very messy. Not sure if there is an acceptable solution.

Reply:
Sounds like there is no elegant fix for adding clang tidy for gmake2. It was fun trying to get it to work but no worries, feel free to close this PR.

Reply:
I succeeded with ninja, and failed with gmake currently (to have a correct per-file per-config)
Some features are already partially working (per-file not taken into account for some generator).
Might be better to have partial working solution than no solution at all...
BTW, I will see if I can improve your PR with my suggestion.

Reply:
Awesome, thanks!